src/Build.UnitTests/Definition/ProjectEvaluationContext_Tests.cs

```diff
@@ -430,10 +430,118 @@ public void ContextCachesExistenceChecksInConditions(EvaluationContext.SharingPo
+                context,
+                project =>
+                {
+                    evaluationCount++;
+
+                    var searchedPath = project.GetProperty("SearchedPath");
+
+                    switch (policy)
+                    {
+                        case EvaluationContext.SharingPolicy.Shared:
+                            searchedPath.EvaluatedValue.ShouldBe(subdirectory.Path);
+                            break;
+                        case EvaluationContext.SharingPolicy.Isolated:
+                            searchedPath.EvaluatedValue.ShouldBe(
+                                evaluationCount == 1
+                                    ? subdirectory.Path
+                                    : _env.DefaultTestDirectory.Path);
+                            break;
+                        default:
+                            throw new ArgumentOutOfRangeException(nameof(policy), policy, null);
+                    }
+
+                    if (evaluationCount == 1)
+                    {
+                        // this will cause the upper file to get picked up in the Isolated policy
+                        subdirectoryFile.Delete();
+                    }
+                });
+
+            evaluationCount.ShouldBe(2);
+        }
+
+        [Theory]
+        [InlineData(EvaluationContext.SharingPolicy.Isolated)]
+        [InlineData(EvaluationContext.SharingPolicy.Shared)]
+        public void ContextCachesExistenceChecksInGetPathOfFileAbove(EvaluationContext.SharingPolicy policy)
+        {
+            var context = EvaluationContext.Create(policy);
+
+            var subdirectory = _env.DefaultTestDirectory.CreateDirectory("subDirectory");
+            var subdirectoryFile = subdirectory.CreateFile("a");
+            var rootFile = _env.DefaultTestDirectory.CreateFile("a");
+
+            int evaluationCount = 0;
+
+            EvaluateProjects(
+                new []
+                {
+                    $@"<Project>
+                      <PropertyGroup>
+                        <SearchedPath>$([MSBuild]::GetPathOfFileAbove('a', '{subdirectory.Path}'))</SearchedPath>
+                      </PropertyGroup>
+                    </Project>"
+                },
+                context,
+                project =>
+                {
+                    evaluationCount++;
+
+                    var searchedPath = project.GetProperty("SearchedPath");
+
+                    switch (policy)
+                    {
+                        case EvaluationContext.SharingPolicy.Shared:
+                            searchedPath.EvaluatedValue.ShouldBe(subdirectoryFile.Path);
+                            break;
+                        case EvaluationContext.SharingPolicy.Isolated:
+                            searchedPath.EvaluatedValue.ShouldBe(
+                                evaluationCount == 1
+                                    ? subdirectoryFile.Path
+                                    : rootFile.Path);
```

Comment:
You've done a few tests like this and it's nice and compact but I'm worried that when a failure comes around it'll take a long time to figure out what the correct behavior was supposed to be. Should we instead have two tests with clearer execution paths?